# Lab 8: Define and Solve an ML Problem of Your Choosing

In [1]:
import pandas as pd
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns

In this lab assignment, you will follow the machine learning life cycle and implement a model to solve a machine learning problem of your choosing. You will select a data set and choose a predictive problem that the data set supports.  You will then inspect the data with your problem in mind and begin to formulate a  project plan. You will then implement the machine learning project plan. 

You will complete the following tasks:

1. Build Your DataFrame
2. Define Your ML Problem
3. Perform exploratory data analysis to understand your data.
4. Define Your Project Plan
5. Implement Your Project Plan:
    * Prepare your data for your model.
    * Fit your model to the training data and evaluate your model.
    * Improve your model's performance.

## Part 1: Build Your DataFrame

You will have the option to choose one of four data sets that you have worked with in this program:

* The "census" data set that contains Census information from 1994: `censusData.csv`
* Airbnb NYC "listings" data set: `airbnbListingsData.csv`
* World Happiness Report (WHR) data set: `WHR2018Chapter2OnlineData.csv`
* Book Review data set: `bookReviewsData.csv`

Note that these are variations of the data sets that you have worked with in this program. For example, some do not include some of the preprocessing necessary for specific models. 

#### Load a Data Set and Save it as a Pandas DataFrame

The code cell below contains filenames (path + filename) for each of the four data sets available to you.

<b>Task:</b> In the code cell below, use the same method you have been using to load the data using `pd.read_csv()` and save it to DataFrame `df`. 

You can load each file as a new DataFrame to inspect the data before choosing your data set.

In [2]:
# File names of the four data sets
adultDataSet_filename = os.path.join(os.getcwd(), "data", "censusData.csv")
airbnbDataSet_filename = os.path.join(os.getcwd(), "data", "airbnbListingsData.csv")
WHRDataSet_filename = os.path.join(os.getcwd(), "data", "WHR2018Chapter2OnlineData.csv")
bookReviewDataSet_filename = os.path.join(os.getcwd(), "data", "bookReviewsData.csv")


df = pd.read_csv(adultDataSet_filename, header=0)

df.head(10)

age         workclass  fnlwgt  education  education-num  \
0  39.0         State-gov   77516  Bachelors             13   
1  50.0  Self-emp-not-inc   83311  Bachelors             13   
2  38.0           Private  215646    HS-grad              9   
3  53.0           Private  234721       11th              7   
4  28.0           Private  338409  Bachelors             13   
5  37.0           Private  284582    Masters             14   
6  49.0           Private  160187        9th              5   
7  52.0  Self-emp-not-inc  209642    HS-grad              9   
8  31.0           Private   45781    Masters             14   
9  42.0           Private  159449  Bachelors             13   

          marital-status         occupation   relationship   race  sex_selfID  \
0          Never-married       Adm-clerical  Not-in-family  White  Non-Female   
1     Married-civ-spouse    Exec-managerial        Husband  White  Non-Female   
2               Divorced  Handlers-cleaners  Not-in-family  White  Non-Female   
3     Married-civ-spouse  Handlers-cleaners        Husband  Black  Non-Female   
4     Married-civ-spouse     Prof-specialty           Wife  Black      Female   
5     Married-civ-spouse    Exec-managerial           Wife  White      Female   
6  Married-spouse-absent      Other-service  Not-in-family  Black      Female   
7     Married-civ-spouse    Exec-managerial        Husband  White  Non-Female   
8          Never-married     Prof-specialty  Not-in-family  White      Female   
9     Married-civ-spouse    Exec-managerial        Husband  White  Non-Female   

   capital-gain  capital-loss  hours-per-week native-country income_binary  
0          2174             0            40.0  United-States         <=50K  
1             0             0            13.0  United-States         <=50K  
2             0             0            40.0  United-States         <=50K  
3             0             0            40.0  United-States         <=50K  
4             0             0            40.0           Cuba         <=50K  
5             0             0            40.0  United-States         <=50K  
6             0             0            16.0        Jamaica         <=50K  
7             0             0            45.0  United-States          >50K  
8         14084             0            50.0  United-States          >50K  
9          5178             0            40.0  United-States          >50K

## Part 2: Define Your ML Problem

Next you will formulate your ML Problem. In the markdown cell below, answer the following questions:

1. List the data set you have chosen.
2. What will you be predicting? What is the label?
3. Is this a supervised or unsupervised learning problem? Is this a clustering, classification or regression problem? Is it a binary classificaiton or multi-class classifiction problem?
4. What are your features? (note: this list may change after your explore your data)
5. Explain why this is an important problem. In other words, how would a company create value with a model that predicts this label?

I have chosen the census data set and using this, I will be predicting the income which is the label of my ML problem. This is a supervised learning problem that utilizes classification as it is grouping people based on whether they made less than or equal to 50k or if they make more than 50k. My features will be every other column but my label. This is an important problem because it can be used to predict how much one makes so if a bank wants to predict if someone is capable of a loan or credit or if a college wants to predict how much aid one recieves, etc., then this will be useful. 

## Part 3: Understand Your Data

The next step is to perform exploratory data analysis. Inspect and analyze your data set with your machine learning problem in mind. Consider the following as you inspect your data:

1. What data preparation techniques would you like to use? These data preparation techniques may include:

    * addressing missingness, such as replacing missing values with means
    * finding and replacing outliers
    * renaming features and labels
    * finding and replacing outliers
    * performing feature engineering techniques such as one-hot encoding on categorical features
    * selecting appropriate features and removing irrelevant features
    * performing specific data cleaning and preprocessing techniques for an NLP problem
    * addressing class imbalance in your data sample to promote fair AI
    

2. What machine learning model (or models) you would like to use that is suitable for your predictive problem and data?
    * Are there other data preparation techniques that you will need to apply to build a balanced modeling data set for your problem and model? For example, will you need to scale your data?
 
 
3. How will you evaluate and improve the model's performance?
    * Are there specific evaluation metrics and methods that are appropriate for your model?
    

Think of the different techniques you have used to inspect and analyze your data in this course. These include using Pandas to apply data filters, using the Pandas `describe()` method to get insight into key statistics for each column, using the Pandas `dtypes` property to inspect the data type of each column, and using Matplotlib and Seaborn to detect outliers and visualize relationships between features and labels. If you are working on a classification problem, use techniques you have learned to determine if there is class imbalance.

<b>Task</b>: Use the techniques you have learned in this course to inspect and analyze your data. You can import additional packages that you have used in this course that you will need to perform this task.

<b>Note</b>: You can add code cells if needed by going to the <b>Insert</b> menu and clicking on <b>Insert Cell Below</b> in the drop-drown menu.

In [3]:
df['sex_selfID'].unique()

array(['Non-Female', 'Female'], dtype=object)

Address missingness, replace outliers, replace the name of fnlwgt, one-hot encoding of gender, race, workclass, education, occupation, relationship

In [4]:
# THIS IS FOR MISSING VALUES:

# Assuming df is your DataFrame containing the dataset
# Check for missing values
missing_values = df.isnull().sum()
print("Missing Values:")
print(missing_values)

# Replace missing numerical values with mean
df['age'].fillna(df['age'].mean(), inplace=True)
df['hours-per-week'].fillna(df['hours-per-week'].mean(), inplace=True)

# Replace missing categorical values with mode
df['workclass'].fillna(df['workclass'].mode()[0], inplace=True)

# Example: Drop rows with missing values in specific columns
df.dropna(subset=['occupation', 'native-country'], inplace=True)

# Verify if missing values have been handled
print("After handling missing values:")
print(df.isnull().sum())

Missing Values:
age                162
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex_selfID           0
capital-gain         0
capital-loss         0
hours-per-week     325
native-country     583
income_binary        0
dtype: int64
After handling missing values:
age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex_selfID        0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income_binary     0
dtype: int64


In [5]:
# THIS IS FOR OUTLIERS

import scipy.stats as stats

hpw_999 = np.percentile(df['hours-per-week'], 99.9)

df['hours-per-week-win'] = stats.mstats.winsorize(df['hours-per-week'], limits=[0.01, 0.01])
df.drop(columns=['hours-per-week'], inplace=True)
df.head()

age         workclass  fnlwgt  education  education-num  \
0  39.0         State-gov   77516  Bachelors             13   
1  50.0  Self-emp-not-inc   83311  Bachelors             13   
2  38.0           Private  215646    HS-grad              9   
3  53.0           Private  234721       11th              7   
4  28.0           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race  sex_selfID  \
0       Never-married       Adm-clerical  Not-in-family  White  Non-Female   
1  Married-civ-spouse    Exec-managerial        Husband  White  Non-Female   
2            Divorced  Handlers-cleaners  Not-in-family  White  Non-Female   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black  Non-Female   
4  Married-civ-spouse     Prof-specialty           Wife  Black      Female   

   capital-gain  capital-loss native-country income_binary  hours-per-week-win  
0          2174             0  United-States         <=50K                40.0  
1             0             0  United-States         <=50K                13.0  
2             0             0  United-States         <=50K                40.0  
3             0             0  United-States         <=50K                40.0  
4             0             0           Cuba         <=50K                40.0

In [6]:
# THIS IS FOR REPLACING NAMES:

columnDict = {'fnlwgt': 'final-weight'}

df = df.rename(columns=columnDict)
df.head()

age         workclass  final-weight  education  education-num  \
0  39.0         State-gov         77516  Bachelors             13   
1  50.0  Self-emp-not-inc         83311  Bachelors             13   
2  38.0           Private        215646    HS-grad              9   
3  53.0           Private        234721       11th              7   
4  28.0           Private        338409  Bachelors             13   

       marital-status         occupation   relationship   race  sex_selfID  \
0       Never-married       Adm-clerical  Not-in-family  White  Non-Female   
1  Married-civ-spouse    Exec-managerial        Husband  White  Non-Female   
2            Divorced  Handlers-cleaners  Not-in-family  White  Non-Female   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black  Non-Female   
4  Married-civ-spouse     Prof-specialty           Wife  Black      Female   

   capital-gain  capital-loss native-country income_binary  hours-per-week-win  
0          2174             0  United-States         <=50K                40.0  
1             0             0  United-States         <=50K                13.0  
2             0             0  United-States         <=50K                40.0  
3             0             0  United-States         <=50K                40.0  
4             0             0           Cuba         <=50K                40.0

In [7]:
# THIS IS FOR ONE-HOT ENCODING

to_encode = list(df.select_dtypes(include=['object']).columns)
to_encode.remove('income_binary')

df[to_encode].nunique()

df_workclass = pd.get_dummies(df['workclass'], prefix='workclass_')
df = df.join(df_workclass)
df.drop(columns = 'workclass', inplace=True)

df_education = pd.get_dummies(df['education'], prefix='education_')
df = df.join(df_education)
df.drop(columns = 'education', inplace=True)

df_marital_status = pd.get_dummies(df['marital-status'], prefix='marital-status_')
df = df.join(df_marital_status)
df.drop(columns = 'marital-status', inplace=True)

df_occupation = pd.get_dummies(df['occupation'], prefix='occupation_')
df = df.join(df_occupation)
df.drop(columns = 'occupation', inplace=True)

df_relationship = pd.get_dummies(df['relationship'], prefix='relationship_')
df = df.join(df_relationship)
df.drop(columns = 'relationship', inplace=True)

df_race = pd.get_dummies(df['race'], prefix='race_')
df = df.join(df_race)
df.drop(columns = 'race', inplace=True)

df_sex_selfID = pd.get_dummies(df['sex_selfID'], prefix='sex_')
df = df.join(df_sex_selfID)
df.drop(columns = 'sex_selfID', inplace=True)

df_country = pd.get_dummies(df['native-country'], prefix='country_')
df = df.join(df_country)
df.drop(columns = 'native-country', inplace=True)

df.head()

age  final-weight  education-num  capital-gain  capital-loss  \
0  39.0         77516             13          2174             0   
1  50.0         83311             13             0             0   
2  38.0        215646              9             0             0   
3  53.0        234721              7             0             0   
4  28.0        338409             13             0             0   

  income_binary  hours-per-week-win  workclass__Federal-gov  \
0         <=50K                40.0                       0   
1         <=50K                13.0                       0   
2         <=50K                40.0                       0   
3         <=50K                40.0                       0   
4         <=50K                40.0                       0   

   workclass__Local-gov  workclass__Private  ...  country__Portugal  \
0                     0                   0  ...                  0   
1                     0                   0  ...                  0   
2                     0                   1  ...                  0   
3                     0                   1  ...                  0   
4                     0                   1  ...                  0   

   country__Puerto-Rico  country__Scotland  country__South  country__Taiwan  \
0                     0                  0               0                0   
1                     0                  0               0                0   
2                     0                  0               0                0   
3                     0                  0               0                0   
4                     0                  0               0                0   

   country__Thailand  country__Trinadad&Tobago  country__United-States  \
0                  0                         0                       1   
1                  0                         0                       1   
2                  0                         0                       1   
3                  0                         0                       1   
4                  0                         0                       0   

   country__Vietnam  country__Yugoslavia  
0                 0                    0  
1                 0                    0  
2                 0                    0  
3                 0                    0  
4                 0                    0  

[5 rows x 105 columns]

## Part 4: Define Your Project Plan

Now that you understand your data, in the markdown cell below, define your plan to implement the remaining phases of the machine learning life cycle (data preparation, modeling, evaluation) to solve your ML problem. Answer the following questions:

* Do you have a new feature list? If so, what are the features that you chose to keep and remove after inspecting the data? 
* Explain different data preparation techniques that you will use to prepare your data for modeling.
* What is your model (or models)?
* Describe your plan to train your model, analyze its performance and then improve the model. That is, describe your model building, validation and selection plan to produce a model that generalizes well to new data. 

For my features I decided to keep all of them because I think that they'll all be important factors to predict the income. I decided that for my data preparation I will address missingness, replace outliers, replace the name of fnlwgt, and one-hot encode gender, race, workclass, education, occupation, and relationship. I already implemented them above. I will use a Logistic Regression model because it is a simple linear model well suited for classification problems such as this one. I'm going to split the data set into its X(label) and y(features), transform the data to improve model performance, and then split them into test and train sets. Then ill fit the model to the logistic regression model and make predictions. To evaluate, i'll test the accuracy score and make a confusion matrix. 

## Part 5: Implement Your Project Plan

<b>Task:</b> In the code cell below, import additional packages that you have used in this course that you will need to implement your project plan.

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score

<b>Task:</b> Use the rest of this notebook to carry out your project plan. 

You will:

1. Prepare your data for your model.
2. Fit your model to the training data and evaluate your model.
3. Improve your model's performance by performing model selection and/or feature selection techniques to find best model for your problem.

Add code cells below and populate the notebook with commentary, code, analyses, results, and figures as you see fit. 

In [10]:
# Sorry i already prepared the data above

# setting up the data
X = df.drop('income_binary', axis=1)
y = df['income_binary']

#splitting the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1234)

# standardizing the data
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# fitting the data to the model
model = LogisticRegression(random_state=42)
model.fit(X_train_scaled, y_train)

# analyzing the model's performance
y_pred = model.predict(X_test_scaled)
print("Accuracy:", accuracy_score(y_test, y_pred))
c_m = confusion_matrix(y_test, y_pred)
pd.DataFrame(
c_m,
columns=['Predicted: Income <= 50k', 'Predicted: Income > 50k'],
index=['Actual: Income <= 50k', 'Actual: Income > 50k']
)

Accuracy: 0.8478981567431375


Predicted: Income <= 50k  Predicted: Income > 50k
Actual: Income <= 50k                      5231                      388
Actual: Income > 50k                        759                     1163

Looking at these results, my model performed pretty well in my opinion as the accuracy score of 84.79% is pretty good and as seen by the confusion matrix, the incorrect predictions are pretty small compared to the actual prediction numbers. 